File: ruby/single-LL.rb
class Node
    attr_accessor :next
    attr_reader   :value
   
    def initialize(value)
      @value = value
      @next  = nil
    end
   
    def to_s
      "Node with value: #{@value}"
    end
  end

class LinkedList
    def initialize
        @head = nil
    end

    def add(value)
        if @head
            find_tail.next = Node.new(value)
        else
            @head = Node.new(value)
        end
    end

    def find_tail
        node=@head
        return node if !node.next
        return node if !node.next while(node=node.next)
    end

    def print
        node = @head
        puts node
        
        while (node = node.next)
            puts node
        end
    end
end

ex = LinkedList.new
ex.add(10)
ex.add(20)
ex.add(30)

ex.print

File: golang/single-LL.go
package main

import "fmt"

type Node struct {
	prev *Node
	next *Node
	key  interface{}
}

type List struct {
	head *Node
	tail *Node
}

func (L *List) Add(key interface{}) {
	list := &Node{
		next: L.head,
		key:  key,
	}
	if L.head != nil {
		L.head.prev = list
	}
	L.head = list

	l := L.head
	for l.next != nil {
		l = l.next
	}
	L.tail = l
}

func (l *List) Show() {
	list := l.head
	for list != nil {
		fmt.Printf("%+v ->", list.key)
		list = list.next
	}
	fmt.Println()
}

func Show(list *Node) {
	for list != nil {
		fmt.Printf("%v ->", list.key)
		list = list.next
	}
	fmt.Println()
}

func ShowBackwards(list *Node) {
	for list != nil {
		fmt.Printf("%v <-", list.key)
		list = list.prev
	}
	fmt.Println()
}

func (l *List) Reverse() {
	curr := l.head
	var prev *Node
	l.tail = l.head

	for curr != nil {
		next := curr.next
		curr.next = prev
		prev = curr
		curr = next
	}
	l.head = prev
	Show(l.head)
}

func main() {
	link := List{}
	link.Add(5)
	link.Add(9)
	link.Add(13)
	link.Add(22)
	link.Add(28)
	link.Add(36)

	fmt.Println("\n==============================\n")
	fmt.Printf("Head: %v\n", link.head.key)
	fmt.Printf("Tail: %v\n", link.tail.key)
	link.Show()
	fmt.Println("\n==============================\n")
	fmt.Printf("head: %v\n", link.head.key)
	fmt.Printf("tail: %v\n", link.tail.key)
	link.Reverse()
	fmt.Println("\n==============================\n")
}
